### PR TITLE
Add rewrite_methods to Redefiner, replacing untyped_methods when type checking is off

### PR DESCRIPTION
**Pull request by obsidiannnn**
## Summary
Implements `rewrite_methods` in `Low::Redefiner` as the new code path when 
`LowType.config.type_checking` is `false`. Foundational piece for low-rb/low_type#38.

## Changes

**`lib/definitions/redefiner.rb`**
- Added `rewrite_methods` - calls `method_proxy.rewrite_signature` to strip type 
  annotations from the shared lines array, then uses `klass.class_eval(method_proxy.export)` 
  to redefine the method directly on the class
- Preserves `private` visibility for methods below `private_start_line`
- `redefine` now accepts optional `klass:` keyword and routes to `rewrite_methods` 
  when type checking is off

**`lib/low_type.rb`**
- Both `redefine` calls now pass `klass:` through
- Result guarded with `if result` since `rewrite_methods` returns `nil`, not a Module

## Depends on
- low-rb/lowkey#9 (`ParamProxy#export(typed: false)` and `MethodProxy#rewrite_signature`)

## Tests
122 examples, 0 failures

## Report

<details>
  <summary>
  Google Summer of Code 2026 — Final Work Report
  </summary>

# Google Summer of Code 2026 — Final Work Report

## Eliminating Shim Methods via Method Rewriting in LowType

**Contributor:** Aditya Chauhan
**Organization:** Ruby
**Mentor:** Maedi
**Program:** Google Summer of Code 2026

---

## 1. Project Overview

The goal of this project was to improve the runtime performance and implementation of method handling in [LowType](https://github.com/low-rb/low_type) by eliminating the need for untyped shim methods when runtime type checking is disabled.

LowType introduces runtime type expressions into Ruby method definitions. When type checking is disabled, the existing implementation uses untyped shim methods. Although this preserves the desired behavior, the additional shim layer can introduce unnecessary runtime overhead.

The primary objective of this project was therefore to implement a method rewriting approach that removes the type annotations from the original method definition and evaluates the resulting Ruby method directly in the target class.

The project required supporting changes in both LowType and its companion [Lowkey](https://github.com/low-rb/lowkey) library. These changes addressed method signature reconstruction, parameter handling, class bindings, constant resolution, type validation, and related performance issues.

---

# 2. Project Goals

The main goals of the GSoC project were:

1. Investigate the existing untyped shim implementation and its performance characteristics.
2. Design and implement a method rewriting mechanism.
3. Replace untyped shim methods with rewritten methods when type checking is disabled.
4. Preserve Ruby method behavior and visibility during rewriting.
5. Correctly resolve user-defined constants used in type expressions.
6. Capture and preserve the appropriate class binding during class loading.
7. Add the supporting functionality required in Lowkey and LowType.
8. Add comprehensive tests for normal and edge-case method signatures.
9. Benchmark the new implementation against the existing shim implementation and plain Ruby.
10. Document the implementation and its configuration.

---

# 3. Summary of Accomplishments

The project resulted in the following major contributions:

- Added class-binding support to Lowkey's `ClassProxy`.
- Added untyped parameter export support to `ParamProxy`.
- Added method signature rewriting support to `MethodProxy`.
- Fixed `ValueExpression` handling during untyped parameter export.
- Normalized `LOWKEY_UNDEFINED` handling.
- Fixed subclass type validation in LowType.
- Removed repeated `Struct.new` allocation from configuration access.
- Added class-binding capture using `TracePoint :end`.
- Added class-level constant evaluation to the evaluator.
- Implemented `rewrite_methods` in LowType's `Redefiner`.
- Added support for typed, untyped, and rewrite configuration modes.
- Preserved private method visibility during rewriting.
- Added 13 edge-case specifications for the rewriting implementation.
- Benchmarked rewritten methods against the existing shim implementation and plain Ruby.
- Updated project documentation.

The supporting changes have been merged according to the current project audit. The primary GSoC deliverable, `rewrite_methods`, is implemented in [LowType PR #45](https://github.com/low-rb/low_type/pull/45) and is currently under review.

---

# 4. Contributions to Lowkey

The method rewriting implementation required several changes to Lowkey.

## PR #6 — Add `class_binding` to `ClassProxy`

**[Lowkey PR #6](https://github.com/low-rb/lowkey/pull/6)**
**Status: Merged**

Added `class_binding` as an `attr_accessor` to `ClassProxy`, initialized to `nil`.

This allows LowType to retain the binding associated with the class body after the class has been loaded.

The captured class binding is required for correctly evaluating user-defined constants in the context in which the class was originally defined.

### Why this was needed

Ruby constant resolution depends on the surrounding class/module context. A generic evaluation context is therefore insufficient for all type-expression cases.

Storing the original class binding provides LowType with the correct context for later evaluation.

---

## PR #9 — Add `export(typed: false)` and `rewrite_signature`

**[Lowkey PR #9](https://github.com/low-rb/lowkey/pull/9)**
**Status: Merged**

Added:

```ruby
ParamProxy#export(typed: false)
```

This allows a parameter to be exported without its LowType type annotation.

The PR also added:

```ruby
MethodProxy#rewrite_signature
```

This reconstructs a method signature using untyped parameter exports and replaces the appropriate signature line in the shared source representation.

### Purpose

This functionality forms the foundation of method rewriting by allowing LowType's type information to be removed while retaining the original Ruby method signature.

---

## PR #10 — Fix `ValueExpression` unwrapping

**[Lowkey PR #10](https://github.com/low-rb/lowkey/pull/10)**
**Status: Merged**

Fixed handling of `ValueExpression` objects during:

```ruby
ParamProxy#export(typed: false)
```

The implementation uses:

```ruby
defined?(ValueExpression) ? ValueExpression : nil
```

to safely check if `ValueExpression` is loaded, then uses `instance_of?` for the actual check. This avoids a `NameError` when LowType is not loaded in Lowkey's context.

Two specifications were added covering:

- positional parameters
- keyword parameters

This ensures that default values represented by `ValueExpression` are correctly unwrapped when generating rewritten method signatures.

---

## PR #11 — Normalize `LOWKEY_UNDEFINED`

**[Lowkey PR #11](https://github.com/low-rb/lowkey/pull/11)**
**Status: Merged**

Fixed `proxy_factory.rb` assigning:

```ruby
':LOWKEY_UNDEFINED'
```

as a string instead of:

```ruby
:LOWKEY_UNDEFINED
```

as a symbol.

The corresponding guards in `param_proxy.rb` were simplified by removing support for both representations in:

- `required?`
- `resolved_default`

This normalized the internal representation and simplified the surrounding implementation.

---

# 5. Contributions to LowType

## PR #29 — Replace strict class equality with `is_a?`

**[LowType PR #29](https://github.com/low-rb/low_type/pull/29)**
**Status: Merged**

### Problem

Type validation used strict class equality:

```ruby
type == value.class
```

This incorrectly rejected instances of subclasses.

### Solution

The validation was changed to:

```ruby
value.is_a?(type)
```

This follows Ruby's inheritance semantics and accepts instances of the declared type and its subclasses.

### Testing

Test coverage was added for:

- direct subclasses
- multi-level inheritance
- core Ruby class hierarchies

### Result

LowType now correctly handles subtype relationships during runtime type validation.

---

## PR #35 — Fix repeated `Struct.new` allocation

**[LowType PR #35](https://github.com/low-rb/low_type/pull/35)**
**Status: Merged**

### Problem

A `Struct.new` definition was being created inside the `config` method.

This caused an anonymous Struct class to be allocated each time the configuration was accessed.

### Solution

The Struct definition was moved out of the method and stored as a constant.

### Result

Repeated calls to `LowType.config` no longer create a new anonymous Struct class, eliminating unnecessary allocation overhead.

---

## PR #39 — Capture class binding using `TracePoint :end`

**[LowType PR #39](https://github.com/low-rb/low_type/pull/39)**
**Status: Merged**

### Problem

LowType needed access to the binding in which a class body had been evaluated.

Without this context, user-defined constants used by type expressions could result in `NameError`.

### Solution

The class-loading phase was wrapped using `TracePoint :end`.

The implementation captures:

```ruby
trace.binding
```

and stores it on:

```ruby
class_proxy.class_binding
```

### Result

LowType can retain the original class body binding and use it later when evaluating expressions.

This became a foundational part of user-defined constant resolution.

---

## PR #40 — Add `class_evaluate` to the Evaluator

**[LowType PR #40](https://github.com/low-rb/low_type/pull/40)**
**Status: Merged**

### Problem

LowType already provided instance-level evaluation, but evaluation in the original class context required a corresponding class-level mechanism.

This caused `NameError` failures when user-defined classes were used in type annotations.

### Solution

Added:

```ruby
class_evaluate
```

alongside the existing:

```ruby
instance_evaluate
```

The captured `class_binding` is threaded through the evaluator call chain.

### Testing

A specification was added covering custom user-defined class resolution.

### Result

User-defined constants can now be evaluated in the correct original class context.

---

# 6. Main GSoC Deliverable

## PR #45 — Add `rewrite_methods` to `Redefiner`

**[LowType PR #45](https://github.com/low-rb/low_type/pull/45)**
**Status: Under Review**

This is the primary GSoC deliverable.

The implementation introduces `rewrite_methods` as an alternative to the existing untyped shim approach when runtime type checking is disabled.

---

## 6.1 Configuration Modes

The implementation supports three modes:

```text
true      → typed methods
false     → existing untyped shim behavior
:rewrite  → new method rewriting implementation
```

This allows the existing behavior to remain available while providing an explicit method-rewriting mode.

---

# 7. How Method Rewriting Works

The rewriting process reconstructs the original method signature without LowType type annotations.

The general flow is:

```text
Typed method
     │
     ▼
MethodProxy
     │
     ▼
Remove type annotations
     │
     ▼
rewrite_signature
     │
     ▼
Generate normal Ruby method
     │
     ▼
class_eval(method_proxy.export)
```

The resulting method is evaluated directly in the target class.

This avoids the additional untyped shim layer used by the previous implementation.

---

# 8. Method Visibility

A key requirement was preserving Ruby method visibility.

The rewriting implementation preserves private method visibility rather than unintentionally turning rewritten methods into public methods.

This behavior is explicitly covered by the additional specifications added with PR #45.

---

# 9. Edge-Case Testing

PR #45 adds **13 new edge-case specifications** covering scenarios including:

- private methods
- positional parameters
- keyword parameters
- default values
- `ValueExpression` unwrapping
- class methods
- methods without parameters

These tests verify that method rewriting works across different Ruby method signatures while preserving expected Ruby semantics.

---

# 10. Issues Opened and Resolved

During the project, the following issues were identified and addressed:

| Issue                                               | Description                              | Status / Resolution                             |
| --------------------------------------------------- | ---------------------------------------- | ----------------------------------------------- |
| [#28](https://github.com/low-rb/low_type/issues/28) | Subclass type checking bug               | Fixed by PR #29                                 |
| [#31](https://github.com/low-rb/low_type/issues/31) | `NameError` for user-defined constants   | Fixed by PR #40                                 |
| [#35](https://github.com/low-rb/low_type/issues/35) | `Struct.new` allocation overhead         | Fixed by PR #35                                 |
| [#38](https://github.com/low-rb/low_type/issues/38) | Method rewriting / main GSoC deliverable | Implemented in PR #45                           |
| [#40](https://github.com/low-rb/low_type/issues/40) | Evaluate constants in original binding   | Addressed through class binding/evaluation work |

---

# 11. Performance Evaluation

The new method rewriting implementation was benchmarked against:

1. Plain Ruby
2. `rewrite_methods`
3. The existing `untyped_methods` shim

The benchmarks were performed using **Ruby 3.3.0**.

---

## 11.1 Keyword Arguments

Keyword arguments represent the primary use case in which the existing shim implementation introduced significant overhead.

| Approach               | Iterations/sec | Relative Result              |
| ---------------------- | -------------: | ---------------------------- |
| Plain Ruby             |      6,200,000 | Baseline                     |
| `rewrite_methods`      |      5,130,000 | 1.21× slower than plain Ruby |
| `untyped_methods` shim |      1,100,000 | 5.66× slower than plain Ruby |

The new rewriting implementation therefore achieved approximately:

### **4.67× higher throughput than the existing shim**

This represents a substantial reduction in the overhead associated with the shim approach.

---

## 11.2 Positional Arguments

| Approach               | Iterations/sec | Result                   |
| ---------------------- | -------------: | ------------------------ |
| Plain Ruby             |     12,600,000 | Baseline                 |
| `rewrite_methods`      |     12,490,000 | Statistically equivalent |
| `untyped_methods` shim |     12,620,000 | Statistically equivalent |

For positional arguments, the rewritten implementation performed at essentially the same level as plain Ruby in the tested benchmark.

---

## 11.3 Object-Reuse Benchmark

The initial keyword-argument benchmark showed a remaining performance gap between rewritten methods and plain Ruby.

Further investigation showed that object allocation inside the benchmark loop contributed to the observed difference.

When the same instance was reused, the results were approximately:

```text
rewrite_methods: 4.80M i/s
plain Ruby:      4.80M i/s
```

Under this configuration, rewritten methods and plain Ruby were statistically identical.

This provides additional evidence that the rewriting mechanism itself does not introduce a significant intrinsic runtime penalty in the tested scenario.

---

# 12. Testing

The reported test suites produced the following results:

| Suite   | Examples | Failures |
| ------- | -------: | -------: |
| Lowkey  |       31 |        0 |
| LowType |      139 |        0 |

### Total

**170 examples — 0 failures**

In addition to the existing test suite, targeted tests were added for the bugs and edge cases encountered during the implementation.

---

# 13. Setup and Reproduction

To clone the repos and run the full test suites:

```bash
# Lowkey
git clone https://github.com/low-rb/lowkey
cd lowkey
bundle install
bundle exec rake spec

# LowType
git clone https://github.com/low-rb/low_type
cd low_type
bundle install
bundle exec rake spec
```

To run the benchmark:

```bash
cd low_type
bundle exec ruby benchmarks/rewriter_benchmark.rb
```

---

# 14. Documentation

Documentation was updated as part of the method rewriting work.

Updated documentation includes:

- [`README.md`](https://github.com/low-rb/low_type/blob/main/README.md)
- [`CHANGELOG.md`](https://github.com/low-rb/low_type/blob/main/CHANGELOG.md)
- [`DEVELOPMENT.md`](https://github.com/low-rb/low_type/blob/main/DEVELOPMENT.md)

These changes document the new behavior and provide context for contributors working with the implementation.

---

# 15. Technical Challenges

## 15.1 Correct Constant Resolution

One of the main challenges was resolving user-defined constants used inside type expressions.

A generic evaluation context was insufficient because Ruby constant lookup depends on the class/module context.

The solution required:

1. Capturing the original class binding.
2. Storing it in `ClassProxy`.
3. Passing it through the evaluator.
4. Adding `class_evaluate`.

This resolved the user-defined constant resolution problem.

---

## 15.2 Reconstructing Ruby Method Signatures

Removing type information from a method signature is more complicated than simply deleting type expressions.

Ruby methods can contain:

- positional arguments
- keyword arguments
- default values
- `ValueExpression` objects
- class methods
- methods without parameters

The `ParamProxy#export(typed: false)` and `MethodProxy#rewrite_signature` functionality was therefore required to correctly reconstruct valid Ruby method definitions.

---

## 15.3 Preserving Ruby Semantics

The rewritten method must behave like the original method.

In particular, the implementation needed to preserve:

- method visibility
- positional arguments
- keyword arguments
- default values
- class methods
- methods without parameters

This resulted in dedicated edge-case coverage in PR #45.

---

## 15.4 Performance Measurement

The objective was not simply to make the new implementation functional, but to determine whether it actually solved the performance problem.

Benchmarking against both the existing shim and plain Ruby provided a direct comparison.

The results demonstrated a substantial improvement for keyword arguments and near-native performance for the tested positional-argument case.

The object-reuse benchmark also highlighted the importance of controlling allocations when interpreting microbenchmark results.

---

# 16. Current State

## Completed and Merged

The following supporting work has been merged:

### Lowkey

- [PR #6](https://github.com/low-rb/lowkey/pull/6) — `class_binding`
- [PR #9](https://github.com/low-rb/lowkey/pull/9) — untyped parameter export and signature rewriting
- [PR #10](https://github.com/low-rb/lowkey/pull/10) — `ValueExpression` unwrapping
- [PR #11](https://github.com/low-rb/lowkey/pull/11) — `LOWKEY_UNDEFINED` normalization

### LowType

- [PR #29](https://github.com/low-rb/low_type/pull/29) — subclass-aware type validation
- [PR #35](https://github.com/low-rb/low_type/pull/35) — Struct allocation fix
- [PR #39](https://github.com/low-rb/low_type/pull/39) — class binding capture
- [PR #40](https://github.com/low-rb/low_type/pull/40) — class-level constant evaluation

---

### [LowType PR #45](https://github.com/low-rb/low_type/pull/45)

The core `rewrite_methods` implementation is complete and includes:

- method signature rewriting
- untyped parameter export
- direct `class_eval`
- configuration modes
- private method visibility preservation
- edge-case tests
- benchmark results
- documentation updates

The PR is currently under review.

---

# 17. Remaining Work

The following work remains:

## 17.1 Review and Merge PR #45

The primary remaining upstream step is completing review of PR #45 and incorporating any requested changes before merge.

---

## 17.2 Add Explicit `class_method` Metadata

The current implementation uses a regex to detect class methods during signature rewriting. This was suggested by the mentor as a non-blocking improvement. A planned improvement is to add a dedicated `class_method` boolean to `MethodProxy` in Lowkey, which would replace regex-based detection with structured method metadata.

---

## 17.3 End-to-End Testing

The implementation should be tested end-to-end against real Raindeer LowNode files.

This would provide additional validation beyond the focused unit and edge-case specifications.

---

## 17.4 Additional Ruby Edge Cases

Further testing should cover:

- blocks
- `method_missing` interactions
- inherited methods
- prepended methods

These cases can be investigated as the implementation moves toward broader upstream adoption.

---

# 18. Key Outcomes

## 18.1 Method Rewriting Was Implemented

LowType now has a `rewrite_methods` path that removes the need for the additional untyped shim layer.

---

## 18.2 Significant Performance Improvement Over the Existing Shim

For the primary keyword-argument benchmark:

```text
Existing shim:     1.10M i/s
Method rewriting:  5.13M i/s
```

This represents approximately:

**4.67× higher throughput.**

---

## 18.3 Near-Native Performance

The positional-argument benchmark produced:

```text
Plain Ruby:       12.60M i/s
rewrite_methods:  12.49M i/s
```

The results were statistically equivalent in the tested benchmark.

The additional object-reuse keyword benchmark also produced approximately 4.80M i/s for both rewritten methods and plain Ruby.

---

## 18.4 Supporting Infrastructure Was Improved

The project also resolved or improved:

- subclass type validation
- class binding capture
- user-defined constant resolution
- repeated Struct allocation
- parameter value handling
- method signature reconstruction

---

# 19. What I Learned

## Ruby Metaprogramming

The project provided substantial experience with Ruby's:

- `class_eval`
- `TracePoint`
- `Binding`
- method definitions
- method visibility
- inheritance
- constant resolution

Understanding the interaction between these mechanisms was essential to implementing method rewriting correctly.

---

## Runtime Performance Analysis

The project reinforced the importance of measuring actual runtime behavior rather than relying only on implementation assumptions.

Benchmarking the existing shim against the rewritten implementation and plain Ruby made the performance impact measurable.

The object-reuse experiment also demonstrated how benchmark design and allocations can influence observed results.

---

## Open Source Development

The project involved:

- breaking work into reviewable PRs
- identifying and tracking issues
- writing tests alongside implementation changes
- responding to implementation problems
- documenting changes
- benchmarking proposed improvements
- iterating toward an upstream-compatible solution

---

## Debugging Through Underlying Behavior

Several issues required investigation beyond the immediate error.

The class-binding and constant-resolution work was particularly valuable in understanding how Ruby's execution context affects metaprogramming and runtime evaluation.

---

# 20. Conclusion

The GSoC 2026 project successfully implemented the core method rewriting approach intended to eliminate the additional untyped shim layer in LowType.

The work produced a new `rewrite_methods` mode, supporting infrastructure in Lowkey and LowType, extensive testing, performance evaluation, and documentation.

The benchmark results demonstrate a substantial performance improvement over the existing shim implementation for keyword arguments while maintaining performance close to plain Ruby.

The supporting changes have been merged, while the primary GSoC deliverable is currently under review in PR #45.

The remaining work consists primarily of completing the upstream review process, performing broader end-to-end validation, and addressing additional Ruby edge cases identified during development.

The implementation provides a foundation for reducing the runtime overhead associated with untyped method shims while preserving Ruby method behavior and enabling future improvements.

---

# 21. Complete Contribution Index

## Lowkey

| PR                                              | Contribution                                       | Status |
| ----------------------------------------------- | -------------------------------------------------- | ------ |
| [#6](https://github.com/low-rb/lowkey/pull/6)   | Add `class_binding` to `ClassProxy`                | Merged |
| [#9](https://github.com/low-rb/lowkey/pull/9)   | Add `export(typed: false)` and `rewrite_signature` | Merged |
| [#10](https://github.com/low-rb/lowkey/pull/10) | Fix `ValueExpression` unwrapping                   | Merged |
| [#11](https://github.com/low-rb/lowkey/pull/11) | Normalize `LOWKEY_UNDEFINED`                       | Merged |

## LowType

| PR                                                | Contribution                                  | Status       |
| ------------------------------------------------- | --------------------------------------------- | ------------ |
| [#29](https://github.com/low-rb/low_type/pull/29) | Replace strict class equality with `is_a?`    | Merged       |
| [#35](https://github.com/low-rb/low_type/pull/35) | Fix repeated `Struct.new` allocation          | Merged       |
| [#39](https://github.com/low-rb/low_type/pull/39) | Capture class binding using `TracePoint :end` | Merged       |
| [#40](https://github.com/low-rb/low_type/pull/40) | Add `class_evaluate`                          | Merged       |
| [#45](https://github.com/low-rb/low_type/pull/45) | Add `rewrite_methods` to `Redefiner`          | Under Review |

## Issues

- [Issue #28](https://github.com/low-rb/low_type/issues/28) — Subclass type checking bug
- [Issue #31](https://github.com/low-rb/low_type/issues/31) — User-defined constant resolution
- [Issue #35](https://github.com/low-rb/low_type/issues/35) — `Struct.new` allocation overhead
- [Issue #38](https://github.com/low-rb/low_type/issues/38) — Method rewriting / GSoC deliverable
- [Issue #40](https://github.com/low-rb/low_type/issues/40) — Original-binding constant evaluation

---

# 22. Final Status

**Project:** Eliminating Shim Methods via Method Rewriting in LowType

**Core implementation:** Implemented

**Supporting infrastructure:** Completed

**Testing:** 170 reported examples, 0 failures

**Performance evaluation:** Completed

**Documentation:** Updated

**Merged upstream:** Supporting Lowkey and LowType contributions listed above

**Primary GSoC deliverable:** [PR #45](https://github.com/low-rb/low_type/pull/45), under review

**Remaining work:** Upstream review/merge, end-to-end validation, and additional edge-case coverage
</details>

### CODE REVIEW

File: spec/fixtures/rewriter.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    count
+  end
+
+  # Required keyword
+  def keyword_required(name: String)
+    name
+  end
+
+  # Optional keyword with default value
+  def keyword_with_default(greeting: String | value('Hello'))
```
Comment (maedi) [inline spec/fixtures/rewriter.rb:26]: `'Hello'` is already a value so it doesn't need to be wrapped in `value()`
